packages/devextreme/js/__internal/ui/html_editor/ui/m_aiDialog.ts

```diff
@@ -0,0 +1,247 @@
+const AI_DIALOG_TITLE_CLASS = 'dx-aidialog-title';
+const AI_DIALOG_TITLE_TEXT_CLASS = 'dx-aidialog-title-text';
+
+const POPUP_MIN_WIDTH = 288;
+const POPUP_MAX_WIDTH = 460;
+const REPLACE_DROPDOWN_WIDTH = 150;
+
+export interface AIDialogShowPayload {
+  currentCommand: string;
+  currentCommandOption?: string;
+  text?: string;
+  commandsMap: CommandsMap;
+  prompt?: HtmlEditorAICustomCommand['prompt'];
+}
+
+export default class AIDialog extends DialogBase {
+  private _isLoading = false;
+
+  private readonly _aIService;
+
+  private _commandsMap: CommandsMap = {};
+
+  private _currentCommand?: string;
+
+  private _currentOption?: string;
+
+  private _commandOptionsList?: string[];
+
+  private _resultText = '';
+
+  private _prompt?: HtmlEditorAICustomCommand['prompt'];
+
+  private _commandSelectBox?: SelectBox;
+
+  private _optionSelectBox?: SelectBox;
+
+  private _resultTextArea?: TextArea;
+
+  private _commandChangeSuppressed = false;
+
+  constructor(editorInstance, aIService?, popupConfig?: PopupProperties) {
+    super(editorInstance, popupConfig);
+
+    this._aIService = aIService;
+  }
+
+  protected _getPopupConfig(): PopupProperties {
+    const baseConfig = super._getPopupConfig();
+
+    return extend(true, {}, baseConfig, {
+      titleTemplate: (titleElement) => {
+        const $titleContainer = $('<div>').addClass(AI_DIALOG_TITLE_CLASS);
+        const $icon = $('<i>').addClass('dx-icon dx-icon-sparkle');
+        const $text = $('<span>')
+          .addClass(AI_DIALOG_TITLE_TEXT_CLASS)
+          .text(localizationMessage.format('dxHtmlEditor-aiDialogTitle'));
+        $titleContainer
+          .append($icon)
+          .append($text);
+        $(titleElement).append($titleContainer);
+      },
+      minWidth: POPUP_MIN_WIDTH,
+      maxWidth: POPUP_MAX_WIDTH,
+      height: 'auto',
+      shading: false,
+      dragEnabled: true,
+      dragAndResizeArea: this._editorInstance?.$element(),
+      toolbarItems: this._getToolbarItems(),
+      showCloseButton: true,
+      hideOnOutsideClick: true,
+      position: {
+        my: 'center',
+        at: 'center',
+        of: this._editorInstance?.$element(),
+      },
+    }) as PopupProperties;
+  }
+
+  protected _renderContent(contentElem: HTMLElement): void {
+    const $container = $(contentElem);
+    $container.addClass(AI_DIALOG_CONTENT_CLASS);
+
+    const $controls = $('<div>')
+      .addClass(AI_DIALOG_CONTROLS_CLASS)
+      .appendTo($container);
+
+    const $commandSelectBox = $('<div>').appendTo($controls);
+    this._commandSelectBox = this._editorInstance._createComponent($commandSelectBox, SelectBox, {
+      value: this._currentCommand,
+      displayExpr: 'text',
+      valueExpr: 'name',
+      onValueChanged: (e) => {
+        if (this._commandChangeSuppressed) {
+          return;
+        }
+
+        this._currentCommand = e.value;
+        this._commandOptionsList = this._commandsMap?.[e.value]?.options ?? [];
```

Comment:
[excess code?]
is really all this stuff nullable? I `_commandsMa[` it's initially `{}`. So it's your util can return nullish value?
<img width="437" alt="image" src="https://github.com/user-attachments/assets/7c5c5e25-e223-4a5c-87ae-f3a2c45b1b37" />